src/main/java/com/ifactory/press/db/solr/HitCount.java

```diff
@@ -1,64 +1,77 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+
+ * http://www.apache.org/licenses/LICENSE-2.0
+
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.ifactory.press.db.solr;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashSet;
+
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.MultiReader;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.queries.function.ValueSource;
+import org.apache.lucene.queries.function.valuesource.DoubleConstValueSource;
+import org.apache.lucene.queries.function.valuesource.SumFloatFunction;
+import org.apache.lucene.queries.function.valuesource.TermFreqValueSource;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.solr.search.FunctionQParser;
+import org.apache.solr.search.SyntaxError;
+import org.apache.solr.search.ValueSourceParser;
+
+/**
+ * Defines the Solr function hitcount([field, ...]) which returns the total
+ * of termfreq(term) for all terms in the query.  The arguments specify
+ * fields whose terms are to be counted.  If no arguments are passed, terms
+ * from every field are counted.
+ */
+public class HitCount extends ValueSourceParser {
+    
+    @Override
+    public ValueSource parse(FunctionQParser fp) throws SyntaxError {
+        // hitcount() takes no arguments.  If we wanted to pass a query
+        // we could call fp.parseNestedQuery()
+        HashSet<String> fields = new HashSet<String>(); 
+        while (fp.hasMoreArguments()) {
+            fields.add(fp.parseArg());
+        }
+        Query q = fp.subQuery(fp.getParams().get("q"), "lucene").getQuery();
+        HashSet<Term> terms = new HashSet<Term>(); 
+        try {
+            /*
+                Lucene 5.1.0 -> 5.2.0 replaced Query.extractTerms with Weight.extractTerms
+                to enforce the requirement that Query.rewrite is supposed to be ran before calling this method.
+                This change's author suggested the following code for Query obj not relying on a specific IndexReader:
+                https://issues.apache.org/jira/browse/LUCENE-6425
+             */
+            IndexReader emptyReader = new MultiReader();
+            new IndexSearcher(emptyReader).createNormalizedWeight(q, true).extractTerms(terms);
+        } catch (UnsupportedOperationException e) {
+            return new DoubleConstValueSource (1);
+        } catch (IOException e) {
+            return new DoubleConstValueSource (1);
+        }
+        ArrayList<ValueSource> termcounts = new ArrayList<ValueSource>();
```

Comment:
Maybe premature optimization, but what about initing this `ArrayList` with the length of `terms`?

Reply:
Good idea. I usually do not do that with ArrayLists if you do not know the definite size, but in this case `terms` is likely to never get very large, so initializing capacity with a little extra wiggle room would not hurt anything